lnrpc/routerrpc/forward_interceptor.go

```diff
@@ -0,0 +1,253 @@
+			log.Errorf("Received an error: %v, shutting down", err)
+			return err
+		case <-s.quit:
+			return nil
+		}
+	}
+}
+
+func (s *Server) setRPCInterceptor(interceptor *rpcInterceptor) error {
+	s.rpcInterceptorMtx.Lock()
+	defer s.rpcInterceptorMtx.Unlock()
+	if s.currentRPCInterceptor != nil {
+		return ErrInterceptorAlreadyExists
+	}
+	rpcInterceptor := interceptor
+	s.currentRPCInterceptor = rpcInterceptor
+	return nil
+}
+
+// onInterceptorDisconnected removes all previousely held forwards from
+// the store. Before they are removed it ensure to resume as the default
+// behavior.
+func (s *Server) onInterceptorDisconnected() {
+
+	log.Infof("RPC interceptor disconnected, resolving held packets")
+	interceptableForwarder := s.cfg.RouterBackend.InterceptableForwarder
+
+	// First de-register the current interceptor so we'll stop getting
+	// more interceptions.
+	interceptableForwarder.SetInterceptor(nil)
+
+	// Now drain and resume all previousely forwards and remove the current
+	// interceptor allows another one to connect.
+	s.rpcInterceptorMtx.Lock()
+	interceptor := s.currentRPCInterceptor
+	s.currentRPCInterceptor = nil
+	s.rpcInterceptorMtx.Unlock()
+
+	interceptor.drain(func(forward htlcswitch.InterceptedForward) {
+		if err := forward.Resume(); err != nil {
+			log.Errorf("failed to resume hold forward %v", err)
+		}
+	})
+}
+
+// rpcInterceptor is a helper struct that handles the lifecycle of an rpc
+// interceptor streaming session.
+// It is created when the stream opens and drains when the stream closes.
+// It should be used from the same go routine and not built for concurrent
+// access.
+type rpcInterceptor struct {
+
+	// holdForwards is a map of current hold forwards and their corresponding
+	// ForwardResolver.
+	holdForwards map[channeldb.CircuitKey]htlcswitch.InterceptedForward
+
+	// stream is the bidirectional stream
+	stream Router_HtlcInterceptorServer
+}
+
+// newRPCInterceptor creates a new rpcInterceptor.
+func newRPCInterceptor(stream Router_HtlcInterceptorServer) *rpcInterceptor {
+	return &rpcInterceptor{
+		stream: stream,
+		holdForwards: make(
+			map[channeldb.CircuitKey]htlcswitch.InterceptedForward),
+	}
+}
+
+// hold just store the intercepted forward.
+func (r *rpcInterceptor) hold(forward htlcswitch.InterceptedForward) {
+	r.holdForwards[forward.CircuitKey()] = forward
+}
+
+// forwardToClient forwards the intercepted htlc to the client.
+func (r *rpcInterceptor) forwardToClient(forward htlcswitch.InterceptedForward) error {
+	htlc := forward.Packet()
+	inKey := forward.CircuitKey()
+
+	// A ChannelAcceptRequest has been received, send it to the client.
+	interceptionRequest := &ForwardHtlcInterceptRequest{
+		CircuitKey: &CircuitKey{
+			IncomingChanId: inKey.ChanID.ToUint64(),
+			HtlcId:         inKey.HtlcID,
+		},
+		HtlcPaymentHash: htlc.PaymentHash[:],
+		AmountSat:       uint64(htlc.Amount.ToSatoshis()),
+		Expiry:          htlc.Expiry,
+	}
+
+	return r.stream.Send(interceptionRequest)
```

Comment:
It feels that the separation of responsibilties in this file `forward_interceptor.go` isn't optimal yet. We are doing sending here, but receive happens outside of the `rpcInterceptor` object.

Reply:
Thanks for the feedback @joostjager. 
In my eyes the main loop is a thin layer that is responsible to consuming input of both sides (switch and the client stream) and delegates the handling of the packets to the `rpcInterceptor` for the actual work. In my mind the "Send" to the stream is equivalent to the "Settle/Fail" on the other side and both are done in the `rpcInterceptor`.
Of course I am opened to suggestions on how to optimize it.

Reply:
I looked at I again and maybe you should just move almost everything from `Server.HtlcInterceptor` into `rpcInterceptor`.

Reply:
Looks good now